packages/salesforcedx-vscode-core/src/commands/templates/libraryBaseTemplateCommand.ts

```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright (c) 2020, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import {
+  TemplateOptions,
+  TemplateService,
+  TemplateType
+} from '@salesforce/templates';
+import {
+  CommandletExecutor,
+  PathStrategyFactory,
+  SelectOutputDir,
+  SourcePathStrategy
+} from '../util';
+
+import { channelService } from '../../channels';
+import { notificationService } from '../../notifications';
+import { telemetryService } from '../../telemetry';
+import {
+  getRootWorkspacePath,
+  hasRootWorkspace,
+  MetadataDictionary,
+  MetadataInfo
+} from '../../util';
+
+import { ContinueResponse } from '@salesforce/salesforcedx-utils-vscode/out/src/types';
+
+import * as path from 'path';
+import { ProgressLocation, window, workspace } from 'vscode';
+
+interface ExecutionResult {
+  output?: string;
+  error?: Error;
+}
+
+/**
+ * Warpping function exeuction with VS Code interface, such as Output channel and progress
+ * @param commandName command name
+ * @param fn function to execute
+ */
+function wrapExecute(
+  commandName: string,
+  fn: (...args: any[]) => Promise<ExecutionResult>
+) {
+  return async (...args: any[]) => {
+    channelService.showCommandWithTimestamp(`Starting ${commandName}`);
+    const result = await window.withProgress(
+      {
+        title: commandName,
+        location: ProgressLocation.Notification
+      },
+      async () => {
+        return await fn.call(null, ...args);
+      }
+    );
+    if (result.output) {
+      channelService.appendLine(result.output);
+      channelService.showCommandWithTimestamp(`Finished ${commandName}`);
+      notificationService.showSuccessfulExecution(commandName).catch(() => {
+        // ignore
+      });
+    }
+    if (result.error) {
+      channelService.appendLine(result.error.message);
+      notificationService.showFailedExecution(commandName);
+    }
+  };
+}
+
+/**
+ * Base class for all template commands
+ */
+export abstract class LibraryBaseTemplateCommand<T>
+  implements CommandletExecutor<T> {
+  private _metadataType: MetadataInfo | undefined;
+  protected showChannelOutput = true;
+
+  /**
+   * Command name
+   */
+  public abstract get executionName(): string;
+  /**
+   * Command telemetry name
+   */
+  public abstract get telemetryName(): string;
+  /**
+   * Template type
+   */
+  public abstract get templateType(): TemplateType;
+  /**
+   * Construct template creation options from user input
+   * @param data data from Continue response
+   */
+  public abstract constructTemplateOptions(data: T): TemplateOptions;
+
+  public async execute(response: ContinueResponse<T>): Promise<void> {
+    const startTime = process.hrtime();
+    await wrapExecute(this.executionName, async () => {
```

Comment:
The idea here is that the call back function is the function in the library you want to execute instead of creating another callback function. That way the wrapExecute is generic enough that it can be used by several library calls.

```
await wrapExecute(this.createTemplate);
...
await this.createTemplate(this.templateType, templateOptions);
```

Here's how we're doing it for the retrieve command https://github.com/forcedotcom/salesforcedx-vscode/blob/810aba681b2cfabf55a9019ff7f4e6910f57ced1/packages/salesforcedx-vscode-core/src/commands/forceSourceRetrieveSourcePath.ts#L127

Reply:
The approach I go with templates each individual command/execute doesn't need to call `createTemplate` explicitly.  It's all done in `LibraryBaseTemplateCommand`. For generalization, `wrapExecute` is designed so that the function it wraps needs to return a raw output to get appended to the output channel, and doesn't require knowledge about business logic.